webhooks/virtualmachine/mutation/cdrom_mutator.go

```diff
@@ -0,0 +1,303 @@
+// number, and controller type. It is used to track available slots for CD-ROM
+// devices on IDE and SATA controllers during mutation.
+type ControllerSlot struct {
+	BusNumber      int32
+	UnitNumber     int32
+	ControllerType vmopv1.VirtualControllerType
+}
+
+// MutateCdromControllerOnUpdate mutates CD-ROM controller assignments for
+// VirtualMachine updates. It automatically assigns controller types, bus
+// numbers, and unit numbers for CD-ROM devices when not specified.
+func MutateCdromControllerOnUpdate(
+	ctx *pkgctx.WebhookRequestContext,
+	_ ctrlclient.Client,
+	vm, _ *vmopv1.VirtualMachine) (bool, error) {
+
+	if !pkgcfg.FromContext(ctx).Features.VMSharedDisks {
+		return false, nil
+	}
+
+	if vm.Spec.Hardware == nil {
+		vm.Spec.Hardware = &vmopv1.VirtualMachineHardwareSpec{}
+	}
+
+	var (
+		mutated     bool
+		hwSpec      = vm.Spec.Hardware
+		usedSlotMap = constructUsedSlotMap(vm)
+	)
+
+	for i := range hwSpec.Cdrom {
+		cdrom := &hwSpec.Cdrom[i]
+		var availableSlot *ControllerSlot
+
+		if cdrom.UnitNumber != nil {
+			if cdrom.ControllerBusNumber != nil && cdrom.ControllerType != "" {
+				ctrlID := pkgutil.ControllerID{
+					ControllerType: cdrom.ControllerType,
+					BusNumber:      *cdrom.ControllerBusNumber,
+				}
+				if addNewController(ctrlID, usedSlotMap, hwSpec) {
+					mutated = true
+				}
+			}
+			// Skip if unit number is set but controller information is
+			// incomplete since we cannot identify the controller.
+			continue
+		}
+
+		// Find available slot based on CD-ROM configuration
+		availableSlot = findAvailableSlot(cdrom, usedSlotMap, hwSpec)
+
+		if availableSlot == nil {
+			continue
+		}
+
+		cdrom.ControllerType = availableSlot.ControllerType
+		cdrom.ControllerBusNumber = &availableSlot.BusNumber
+		cdrom.UnitNumber = &availableSlot.UnitNumber
+
+		ctrlID := pkgutil.ControllerID{
+			ControllerType: cdrom.ControllerType,
+			BusNumber:      *cdrom.ControllerBusNumber,
+		}
+		usedSlotMap[ctrlID][*cdrom.UnitNumber] = true
+		mutated = true
+	}
+
+	return mutated, nil
+}
+
+// findNextAvailableSATASlot finds the next available slot on a SATA controller.
+// It returns a ControllerSlot with the bus number and unit number, or nil if no
+// slots are available.
+func findNextAvailableSATASlot(
+	usedSlotMap map[pkgutil.ControllerID][]bool,
+	hwSpec *vmopv1.VirtualMachineHardwareSpec) *ControllerSlot {
+
+	var nextBusNum *int32
+
+	for busNum := range pkgutil.SATAControllerMaxCount {
+		ctrlID := pkgutil.ControllerID{
+			ControllerType: vmopv1.VirtualControllerTypeSATA,
+			BusNumber:      busNum,
+		}
+		if _, exists := usedSlotMap[ctrlID]; !exists {
+			if nextBusNum == nil {
+				nextBusNum = &busNum
+			}
+			continue
+		}
+
+		for unitNum := range pkgutil.SATAControllerMaxSlotCount {
+			if !usedSlotMap[ctrlID][unitNum] {
+				return &ControllerSlot{
+					BusNumber:      busNum,
+					UnitNumber:     unitNum,
+					ControllerType: vmopv1.VirtualControllerTypeSATA,
+				}
+			}
+		}
+	}
+
```

Comment:
nit: maybe add a comment that states no available slots for existing buses, so need to create a new one?